crates/store/src/accounts/mod.rs

```diff
@@ -0,0 +1,490 @@
+    fn get(&self, account_id: AccountId) -> Word;
+
+    /// Computes mutations for applying account updates.
+    fn compute_mutations(
+        &self,
+        accounts: impl IntoIterator<Item = (AccountId, Word)>,
+    ) -> Result<AccountMutationSet, AccountTreeError>;
+
+    /// Applies mutations with reversion data.
+    fn apply_mutations_with_reversion(
+        &mut self,
+        mutations: AccountMutationSet,
+    ) -> Result<AccountMutationSet, AccountTreeError>;
+
+    /// Checks if the tree contains an account with the given prefix.
+    fn contains_account_id_prefix(&self, prefix: miden_objects::account::AccountIdPrefix) -> bool;
+}
+
+impl<S> AccountTreeStorage for AccountTree<LargeSmt<S>>
+where
+    S: SmtStorage,
+{
+    fn root(&self) -> Word {
+        self.root()
+    }
+
+    fn num_accounts(&self) -> usize {
+        self.num_accounts()
+    }
+
+    fn open(&self, account_id: AccountId) -> AccountWitness {
+        self.open(account_id)
+    }
+
+    fn get(&self, account_id: AccountId) -> Word {
+        self.get(account_id)
+    }
+
+    fn compute_mutations(
+        &self,
+        accounts: impl IntoIterator<Item = (AccountId, Word)>,
+    ) -> Result<AccountMutationSet, AccountTreeError> {
+        self.compute_mutations(accounts)
+    }
+
+    fn apply_mutations_with_reversion(
+        &mut self,
+        mutations: AccountMutationSet,
+    ) -> Result<AccountMutationSet, AccountTreeError> {
+        self.apply_mutations_with_reversion(mutations)
+    }
+
+    fn contains_account_id_prefix(&self, prefix: miden_objects::account::AccountIdPrefix) -> bool {
+        self.contains_account_id_prefix(prefix)
+    }
+}
+
+// ERROR TYPES
+// ================================================================================================
+
+#[allow(missing_docs)]
+#[derive(thiserror::Error, Debug)]
+pub enum HistoricalError {
+    #[error(transparent)]
+    MerkleError(#[from] MerkleError),
+    #[error(transparent)]
+    AccountTreeError(#[from] AccountTreeError),
+}
+
+// HISTORICAL STATE ENUM
+// ================================================================================================
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum HistoricalState {
+    /// The requested block is in the future (later than current block).
+    Future,
+    /// The requested block is available in history.
+    Target(BlockNumber),
+    /// The requested block is the current/latest block.
+    Latest,
+    /// The requested block is too old and has been pruned from history.
+    TooAncient,
+}
+
+// HISTORICAL OVERLAY
+// ================================================================================================
+
+/// Captures reversion state for historical queries at a specific block.
+#[derive(Debug, Clone)]
+struct HistoricalOverlay {
+    block_number: BlockNumber,
+    root: Word,
+    node_mutations: HashMap<NodeIndex, Word>,
+    account_updates: HashMap<Word, Word>,
```

Comment:
Should we use `HashMap` from `hashbrown` here? IIRC, it is a bit more efficient than the one from Rust stdlib.